lib/msf/core/exploit/remote/http/exchange.rb

```diff
@@ -0,0 +1,122 @@
+          @@exchange_server_2019_builds = ['15.2.986.15', '15.2.986.14', '15.2.986.9', '15.2.986.5', '15.2.922.20', '15.2.922.19', '15.2.922.14', '15.2.922.13', '15.2.922.7', '15.2.858.15', '15.2.858.12', '15.2.858.10', '15.2.858.5', '15.2.792.15', '15.2.792.13', '15.2.792.10', '15.2.792.3', '15.2.721.13', '15.2.721.2', '15.2.659.12', '15.2.659.4', '15.2.595.8', '15.2.595.3', '15.2.529.13', '15.2.529.5', '15.2.464.15', '15.2.464.5', '15.2.397.11', '15.2.397.3', '15.2.330.11', '15.2.330.5', '15.2.221.18', '15.2.221.12', '15.2.196.0']
+
+          # Full list of all known Exchange Server builds.
+          @@exchange_builds = [*@@exchange_server_4_0_builds, *@@exchange_server_5_0_builds, *@@exchange_server_5_5_builds, *@@exchange_server_2000_builds, *@@exchange_server_2003_builds, *@@exchange_server_2007_builds, *@@exchange_server_2010_builds, *@@exchange_server_2013_builds, *@@exchange_server_2016_builds, *@@exchange_server_2019_builds]
+
+          # Get the Exchange version number.
+          #
+          # @see https://docs.microsoft.com/en-us/exchange/new-features/build-numbers-and-release-dates Exchange Version Numbers
+          #
+          # @return [Rex::Version, nil] The Exchange version if it was able to be recovered. Nil otherwise
+          def exchange_get_version
+            # First lets try a cheap way of doing this via a leak of the X-OWA-Version header.
+            # If we get this we know the version number for sure and we can skip a lot of leg work.
+            res = send_request_cgi(
+              'method' => 'GET',
+              'uri' => normalize_uri(target_uri.path, '/owa/service')
+            )
+
+            unless res
+              print_error('Target did not respond!')
+              return nil
+            end
+
+            if res.headers['X-OWA-Version']
+              build = res.headers['X-OWA-Version']
+              return Rex::Version.new(build)
+            end
+
+            # Next, determine if we are up against an older version of Exchange Server where
+            # the /owa/auth/logon.aspx page gives the full version. Recent versions of Exchange
+            # give only a partial version without the build number.
+            res = send_request_cgi(
+              'method' => 'GET',
+              'uri' => normalize_uri(target_uri.path, '/owa/auth/logon.aspx')
+            )
+
+            unless res
+              print_error('Target did not respond!')
+              return nil
+            end
+
+            if res.code == 200 && ((%r{/owa/(?<build>\d+\.\d+\.\d+\.\d+)} =~ res.body) || (%r{/owa/auth/(?<build>\d+\.\d+\.\d+\.\d+)} =~ res.body))
```

Comment:
Pretty sure the expression can be simplified here:

```
if res.code == 200 && res.body =~ %r{/owa/(auth/)?(?<build>\d+(\.\d+){3})}
```

It's mostly removing the second regex.

Reply:
Looks good to me, thanks for the suggestion, will implement that now!

Reply:
Suggested fix implemented in https://github.com/rapid7/metasploit-framework/pull/16249/commits/08032259599322475fb8d7db28c0c4e043c49fdc